WARNING pragmas with categories (#541)
ghc-proposals/ghc-proposals: proposals/0000-warning-pragmas-with-categories.rst

============  =  =====================================================================
*decl*        →  ``{-# WARNING`` [*categories*] *things* *strings* ``#-}``
*decl*        →  ``{-# DEPRECATED`` [*categories*] *things* *strings* ``#-}``

*categories*  →  ``[`` *category1*, ..., *categoryN* ``]``
*category*    →  (*small* | *large*) { *small* | *large* | *digit* | ``'`` | ``-`` }

*things*      →  *thing1*, ..., *thingN*
*thing*       →  *varid* | *conid*
*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
============  =  =====================================================================

The list of categories can be omitted entirely, so this subsumes the existing
syntax for ``WARNING`` and ``DEPRECATED`` pragmas.  That is, [*categories*]
means that the presence of the *categories* non-terminal is optional; if present
it is a list in square brackets (hence the use of typewriter font ``[``
... ``]`` in the definition of the *categories* production).

The *category* non-terminal subsumes both *varid* and *conid*, so it is always
possible to use the name of the thing to which a warning is being attached as

> one of the categories.  The dash character (``-``) is permitted as a character

Reviewer: Neat idea, but do we really want tools that deal with GHC command lines deal with all that complexity? Unicode emojis etc?

Also, a name can be many things. Can I use fully qualified names? Maybe even package qualified names?

Also, multiple categories is something we don't have yet. Does that mean it matters in which order I pass `-Werror=` or `-Wno-` etc. 

I'd aim for a single category only consisting of lower case letters, dashes, and maybe numbers, and keep it simple.

Author: > Neat idea, but do we really want tools that deal with GHC command lines deal with all that complexity? Unicode emojis etc?

Fair question. I don't have much insight into the details of cross-platform CLI handling, so I don't know what an appropriate set of restrictions on characters would be. (Obviously one could permit only ASCII lowercase letters, digits, and dashes, but might that be an annoying for restriction for users in some locales?)

One nice aspect of the present proposal is that every Haskell identifier name is a valid warning category, so library authors can choose to have the category name match the identifier name. That's also why I suggested allowing categories to start with uppercase letters. But I don't hugely care about the exact choices here.

> Also, a name can be many things. Can I use fully qualified names? Maybe even package qualified names?

No, this is mentioned briefly in the Costs and Drawbacks section.

> Also, multiple categories is something we don't have yet. Does that mean it matters in which order I pass -Werror= or -Wno- etc. (Answer below, I know, but I wish I would not have to ask :-))

Perhaps we don't directly have multiple categories, but we do have warning groups (e.g. `-W`, `-Wcompat`, `-Wall`) which amount to much the same thing, just with a restricted list of groups. In particular, we already have order-dependence (e.g. `-Wno-missing-signatures -Wall` is equivalent to `-Wall` and not `-Wall -Wno-missing-signatures`).

I think the "tagging" approach that multiple categories permits is a simple way to get behaviour like the existing warning groups, without needing a predefined taxonomy.

in addition to identifier characters, since dashes are frequently used in
warning names.

There is no change to the existing rules for when ``WARNING`` or ``DEPRECATED``
pragmas give rise to warnings, except that individual warning categories may be
enabled or disabled using new ``-Wwarnings:<category>`` or
``-Wno-warnings:<category>`` options, and their priority may be controlled using
the ``-Werror=warnings:<category>`` or ``-Wwarn=warnings:<category>`` options.
Here ``<category>`` represents the name of a *category* according to the grammar
above.  The ``-Wwarnings-deprecations`` warning flag (and its synonym
``-Wdeprecations``) controls the display of all warnings from ``WARNING`` or

> ``DEPRECATED`` pragmas, regardless of category.  The command-line flags are

Reviewer: I'd have expected that no category now defaults to “`deprecation`”, and that flag still affects only that category - preserves the existing semantics, without making a flag with deprecation in the name disabling unrelated warnings (that happen to be specified with a related tool)

Author: Hmm, I quite like this idea. In particular I think we could say:
 * `{- # WARNING ... #-}` or `{- # DEPRECATED ... #-}` without a category list is a synonym for `{-# WARNING [deprecation] ... #-}` ( and perhaps `DEPRECATED` cannot have a category list)
 * `-Wwarnings` (new) enables all warnings from `WARNING` pragmas regardless of category
 * `-Wwarnings-deprecations` and `-Wdeprecations` become synonyms for `-Wwarnings:deprecation`

I'd vaguely worried about backwards compatibility here, but I guess the only potential surprise is that if a dependency starts using a new category, a downstream library that sets `-Wno-warnings-deprecations` or `-Wno-deprecations` will start seeing the warning until it sets -`Wno-warnings`. That seems okay.

Author: Strictly speaking we could do without it. But we have `-Wwarnings-deprecations` at the moment, and given all the flexibility to configure warnings it would seem odd if there wasn't a way to say "don't show me any messages from `WARNING` pragmas" without disabling all built-in warnings as well.

Reviewer: I'd say users don't care too much if a warning comes from a pragma or some other source (built in? Plugin?), and would design the feature from that point of view. 

But it's not a big deal to add this warning category group, either, if you think it's useful.

type-level programming to control the presence and content of warnings, and the
suppression of warnings at individual use sites.

However, this proposal is simpler, and by keeping the ``WARNING`` annotations as
separate pragmas rather than requiring them to be part of the types, avoids the
risk that introducing ``Warning`` constraints may have unexpected effects on
program semantics.  Assuming this proposal is accepted, it would be fairly
simple to change the ``Warning`` class to be controlled using the same flags.

`Proposal #454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_ does
not allow multiple categories to be attached to a single warning, but it seems
useful to include this feature, because it allows for multiple categories at
different levels of granularity (potentially including a different category for
every identifier).

Unresolved Questions
--------------------

GHC currently supports both ``WARNING`` and ``DEPRECATED`` pragmas, which are
synonymous.  Similarly the flags ``-Wwarnings-deprecations`` and
``-Wdeprecations`` are synonymous, and there is no ``-Wwarnings`` flag.  This
seems somewhat inconsistent.  Should we reduce the redundancy by considering

> deprecations one particular category of warnings?

Reviewer: I'd say so

Reviewer: And `-Wwarnings-deprecations` and `-Wdeprecations` simply become synonyms for `-Wwarnings:deprecation` then.

Could the use of a colon in the command-line flag name cause issues on some

> platforms?  Is there another syntax that would be preferable?

Reviewer: I considered injecting the custom categories into the space of all categories via prefix `ext-` or `x-`. Concise, but a bit idiosyncratic.

Author: I'd also previously considered `custom:` but I quite like `x-`. I suggested an explicit `:` or such to make clear the distinction between parts of the flag name that were determined by user code, but perhaps that's unnecessary.

Reviewer: The end user doesn't care too much about where the warning came from originally (built in or custom), so the least intrusive variant is best.

I'd have suggested to not mangle the name at all, allowing library authors to add warnings to existing built-in categories! The only reason I don't is that this would make it hard for GHC to complain about unrecognized warning categories.

emitted when the functions are used.  It has been controversial because, while
there is broad consensus that partial functions should *usually* be avoided,
there is less consensus that they should *never* be used.

A key problem with ``WARNING`` pragmas as they currently exist is that there is
no way to disable them selectively. It is possible to disable them entirely,
with ``-Wno-warnings-deprecations``, but this will suppress all warnings from
``WARNING`` pragmas, regardless of source.  Thus if the Core Libraries Committee
decide to add a ``WARNING`` to partial functions, users who wish to continue
using them will either need to put up with warnings that they consider
unhelpful, suppress potentially-helpful warnings, or define their own
non-standard versions of the functions.  This makes the change more
controversial than it should be, as it potentially inconveniences existing
users.

Instead, this ghc-proposal extends ``WARNING`` pragmas so they can be annotated
with a "warning category".  Users can then enable or disable these
categories individually.  For example, it would allow the core-libraries
proposal to introduce::

>     {-# WARNING [partial] head "This is a partial function, it throws an error on empty lists." #-}
>     {-# WARNING [partial] tail "This is a partial function, it throws an error on empty lists." #-}

Reviewer: ```suggestion
    {-# WARNING [x-partial] head "This is a partial function, it throws an error on empty lists." #-}
    {-# WARNING [x-partial] tail "This is a partial function, it throws an error on empty lists." #-}
```

square brackets. A warning category is an identifier beginning with a letter
and followed by valid identifier characters or dashes.

More formally, the grammar of declarations is extended as follows:

============  =  =====================================================================
*decl*        →  ``{-# WARNING`` [*categories*] [*things*] *strings* ``#-}``

*categories*  →  ``[`` *category* ``]``
*category*    →  (*small* | *large*) { *small* | *large* | *digit* | ``'`` | ``-`` }

*things*      →  *thing1*, ..., *thingN*
*thing*       →  *varid* | *conid*
*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
============  =  =====================================================================

The category can be omitted entirely, so this subsumes the existing
syntax for ``WARNING`` pragmas.  That is, [*categories*]
means that the presence of the *categories* non-terminal is optional; if present
it is a single category in square brackets (hence the use of typewriter font ``[``
... ``]`` in the definition of the *categories* production).
Similarly the list of *things* is optional as it may be omitted (for a ``WARNING`` on a module header).

> The *category* non-terminal subsumes both *varid* and *conid*, so it is
> possible to use the name of the thing to which a warning is being attached as
> the category, provided it is not an operator.  The dash character (``-``) is permitted as a character
> in addition to identifier characters, since dashes are frequently used in
> warning names.
> 

Reviewer: ```suggestion
The *category* non-terminal subsumes both *varid* and *conid*, so it is
possible to use the name of the thing to which a warning is being attached as
the category, provided it is not an operator.  The dash character (``-``) is permitted as a character
in addition to identifier characters, since dashes are frequently used in
warning names.

The compiler only allows a *category* that starts with `x-` or is `deprecations`.
(This restriction is not part of the syntax change because we may expect later
extensions to that list.)

```

> There is no change to the existing rules for when ``WARNING``
> pragmas give rise to warnings, except that individual warning categories may be
> enabled or disabled using new ``-Wx-<category>`` or
> ``-Wno-x-<category>`` options, and their priority may be controlled using
> the ``-Werror=x-<category>`` or ``-Wwarn=x-<category>`` options.
> Here ``<category>`` represents the name of a *category* according to the grammar
> above.  The command-line flags are
> processed from left to right, with later flags overriding previous ones, as at
> present.

Reviewer: ```suggestion
There is no change to the existing rules for when ``WARNING``
pragmas give rise to warnings, except that individual warning categories may be
enabled or disabled using new ``-W<category>`` or
``-Wno-<category>`` options, and their priority may be controlled using
the ``-Werror=<category>`` or ``-Wwarn=<category>`` options.
Here ``<category>`` represents the name of a *category* according to the grammar
above.  The command-line flags are
processed from left to right, with later flags overriding previous ones, as at
present.
```

``-Wno-warnings-deprecations``   Warnings displayed (category is not ``deprecations``)
===============================  ===============================================

Warning severity levels may be overridden by subsequent arguments on the
command-line.  For example, ``-Wno-extended-warnings -Werror=x-partial-head``
will result in errors instead of warnings with the category ``partial-head``,
but no other warnings from ``WARNING`` or ``DEPRECATED`` pragmas.  On the other
hand, ``-Werror=x-partial-head -Wno-extended-warnings`` will result in no
warnings because the second option overrides the first.

Effect and Interactions
-----------------------

This proposal should help resolve the controversy over whether ``head`` and
``tail`` should be annotated with ``WARNING`` pragmas.  By annotating them with
categorised warnings, users will be warned about their use by default, but may
choose to override the warnings as they wish.

This approach also provides an alternative to `proposal #528
<https://github.com/ghc-proposals/ghc-proposals/pull/528>`_, which is about
discouraging users from importing from "internal" modules, without completely

> prohibiting their import.  For example, a ``WARNING [ghc-prim-internals]``

Reviewer: ```suggestion
prohibiting their import.  For example, a ``WARNING [x-ghc-prim-internals]``
```

pragma could be attached to all module in ``ghc-prim``.  Users would then be
advised that such imports are discouraged, but could silence the warning with
``-Wno-x-ghc-prim-internals``.

These pragmas may be useful for libraries outside ``base`` as well, in
particular where library authors wish to selectively discourage use of certain
parts of their API.

Costs and Drawbacks
-------------------

This is yet one more feature to implement, although the implementation cost
should be fairly modest.

Overall this should make the language more accessible to newcomers, as library
authors will be able to use ``WARNING`` pragmas to discourage certain features
even if those warnings can be reasonably be disabled in some contexts.

This proposal does not provide a way to disable warnings at specific use sites,
only at the module level.  In some cases, it would be nice to be able to mark
individual uses as having been approved and the warning suppressed for that use
alone, rather than for all uses in the module.

It might be helpful to establish conventions around which categories exist, such

> as ``partial`` for warnings about partial functions.  These issues are currently

Reviewer: ```suggestion
as ``x-partial`` for warnings about partial functions.  These issues are currently
```

*thing*       →  *varid* | *conid*
*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
============  =  =====================================================================

The category can be omitted entirely, so this subsumes the existing
syntax for ``WARNING`` pragmas.  That is, [*categories*]
means that the presence of the *categories* non-terminal is optional; if present
it is a single category in square brackets (hence the use of typewriter font ``[``
... ``]`` in the definition of the *categories* production).
Similarly the list of *things* is optional as it may be omitted (for a ``WARNING`` on a module header).

The *category* non-terminal subsumes both *varid* and *conid*, so it is
possible to use the name of the thing to which a warning is being attached as
the category, provided it is not an operator.  The dash character (``-``) is permitted as a character
in addition to identifier characters, since dashes are frequently used in
warning names.

Examples
--------

Suppose the definitions of ``head`` and ``nub`` are annotated with::

    {-# WARNING [x-partial] head "This is a partial function, it throws an error on empty lists." #-}

>     {-# WARNING [x-quadratic] tail "The nub function has quadratic run-time complexity. If possible, use nubBy or nubOn." #-}

Reviewer 2: Still says `tail`

Reviewer: Thanks, fixed on master now.

   ``deprecations``.  GHC will emit an error if a ``WARNING`` pragma uses a
   category that is not recognised.

There is no change to the existing rules for when ``WARNING`` pragmas give rise
to warnings; these changes merely affect whether the warnings are displayed.
The command-line flags are processed from left to right, with later
flags overriding previous ones, as at present.

Points 3 and 4 make the proposal backwards compatible: existing ``WARNING`` or
``DEPRECATED`` pragmas will still be controlled via the ``-Wdeprecations`` or
``-Wwarnings-deprecations`` options.  However if a library introduces a category
for a previously uncategorised warning, the warning will no longer be suppressed
by ``-Wno-deprecations``.  (Prior to this proposal, all ``WARNING`` and
``DEPRECATED`` pragmas were treated uniformly and could be disabled either by
``-Wno-deprecations`` or by ``-Wno-warnings-deprecations``.)

Point 5 is not strictly necessary, but ``-Wno-extended-warnings`` will allow
users to suppress all ``WARNING`` messages much like
``-Wno-warnings-deprecations`` does at present.  The naming is chosen to allow
for other sources of "extended warnings" in the future.

> The restriction to recognised categories in point 6 means that ``WARNING``
> pragmas cannot extend the behaviour of existing GHC warning flags such as
> ``-Wtabs`` or groups such as ``-Wall``.  Moreover it means that GHC can still
> report unrecognised warning flags, rather than silently accepting them.  In the
> future we may extend the set of recognised categories, if it appears useful to
> let users extend specific built-in warning categories (e.g. ``-Wcompat``) or to
> allow prefixes other than ``x-``.

Reviewer: ```suggestion
Point 6 makes sure that the command line options `-Wx-partial` and `-Wno-x-partial`
can readily be distinguished from the existing large number of built-in warning
categories, such as `-Wtabs`, `-Wdodgy-import`, `-Winaccessible-code` etc. (See the
`user manual section <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/using-warnings.html?highlight=warning#warnings-and-sanity-checking>`_ for a complete list).
This way GHC can still
report unrecognised warning flags, rather than silently accepting them.
```
